server/utils/auth.py

```diff
@@ -0,0 +1,117 @@
+from os import abort
+from flask import redirect, request, session, url_for, render_template
+from flask_login import LoginManager, login_user, logout_user, login_required
+from flask_oauthlib.client import OAuth
+from oauth2client.contrib.flask_util import UserOAuth2
+
+
+from server import app
+from server.models import db, User
+import server.utils.canvas as canvas_client
+from server.utils.stub import get_dev_user_oauth_resp
+from server.form import DevLoginForm
+
+login_manager = LoginManager(app=app)
+
+oauth = OAuth()
+
+server_url = app.config.get('CANVAS_SERVER_URL')
+consumer_key = app.config.get('CANVAS_CLIENT_ID')
+consumer_secret = app.config.get('CANVAS_CLIENT_SECRET')
+
+canvas_oauth = oauth.remote_app(
+    'seating',
+    consumer_key=consumer_key,
+    consumer_secret=consumer_secret,
+    base_url=server_url,
+    request_token_url=None,
+    access_token_method='POST',
+    access_token_url=server_url + 'login/oauth2/token',
+    authorize_url=server_url + 'login/oauth2/auth',
+)
+
+
+@canvas_oauth.tokengetter
+def get_access_token(token=None):
+    return session.get('access_token')
+
+
+google_oauth = UserOAuth2(app)
+
+
+@login_manager.user_loader
+def load_user(user_id):
+    return User.query.get(user_id)
+
+
+@login_manager.unauthorized_handler
+def unauthorized():
+    session['after_login'] = request.url
+    return redirect(url_for('login'))
+
+
+@app.route('/login/')
+def login():
+    if canvas_client.is_mock_canvas():
+        return redirect(url_for('dev_login'))
+    return canvas_oauth.authorize(callback=url_for('authorized', _external=True, _scheme='https'))
+
+
+@app.route('/dev_login/', methods=['GET', 'POST'])
+def dev_login():
+    if canvas_client.is_mock_canvas():
+        form = DevLoginForm()
+        if form.validate_on_submit():
+            if form.login_as_jimmy.data:
+                session["dev_user_id"] = 234567
+                return redirect(url_for('authorized'))
+            elif form.login_as_yu.data:
+                session["dev_user_id"] = 123456
+                return redirect(url_for('authorized'))
+            else:
+                abort(500, 'Invalid dev user')
+        return render_template('dev_login.html.j2', form=form)
+    return redirect(url_for('index'))
+
+
+@app.route('/authorized/')
+def authorized():
+    if canvas_client.is_mock_canvas():
+        resp = get_dev_user_oauth_resp(session["dev_user_id"])
```

Comment:
This suggests the mock canvas could be its own subclass of the CanvasClient which could reduce some conditional logic